core/src/err.rs

```diff
@@ -188,3 +252,28 @@ impl Error {
         }
     }
 }
+
+impl ErrorDetails for Error {
+    fn details(&self) -> Option<&QueryDiagnostics> {
+        match self {
+            Error::InvalidCursorStateWithDiagnostics(diagnostics)
+            | Error::ColIndexOutOfBoundsWithDiagnostics(_, diagnostics)
+            | Error::ValueAccessWithDiagnostics(_, _, diagnostics)
+            | Error::QueryCursorUpdateWithDiagnostics(_, diagnostics) => Some(diagnostics),
+            _ => None,
+        }
+    }
+}
+
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_error_details_exist() {
+        let diagnostics = QueryDiagnostics {
+            query: Arc::from("test_query"),
+            schema: Arc::from("test_schema"),
+            translation: Arc::from("test_translation"),
+        };
+    }
+}
```

Comment:
Determining the best way to have a test that ensures errors that should have diagnostics do have diagnostics so we can catch it during testing rather than an `assert` at runtime.